conans/client/downloaders/caching_file_downloader.py

```diff
@@ -43,17 +47,49 @@ def _lock(self, lock_id):
             finally:
                 thread_lock.release()
 
-    def _caching_download(self, url, file_path, md5, sha1, sha256, **kwargs):
-        h = self._get_hash(url, md5, sha1, sha256)
+    def _caching_download(self, url, file_path, md5, sha1, sha256, conanfile: ConanFile, **kwargs):
+        sources_cache = False
+        h = None
+        if conanfile is not None:
+            if sha256:
+                h = sha256
+                sources_cache = True
+            else:
+                ConanOutput()\
+                    .warning("Expected sha256 to be used as file checksums for downloaded sources")
+        if h is None:
+            h = self._get_hash(url, md5, sha1, sha256)
+
         with self._lock(h):
-            cached_path = os.path.join(self._download_cache, h)
+            if sources_cache:
+                cached_path = os.path.join(self._download_cache, 's', h)
+            else:
+                cached_path = os.path.join(self._download_cache, 'c', h)
             remove_if_dirty(cached_path)
 
             if not os.path.exists(cached_path):
                 with set_dirty_context_manager(cached_path):
-                    self._file_downloader.download(url=url, file_path=cached_path, md5=md5,
+                    self._file_downloader.download(url, cached_path, md5=md5,
                                                    sha1=sha1, sha256=sha256, **kwargs)
+            if sources_cache:
+                summary_path = cached_path + ".json"
+                if os.path.exists(summary_path):
+                    summary = json.loads(load(summary_path))
+                else:
+                    summary = {}
+
+                try:
+                    summary_key = conanfile.ref.repr_notime()
+                except AttributeError:
+                    if conanfile.recipe_folder is not None:
+                        summary_key = conanfile.recipe_folder
```

Comment:
Recipe folder might be also variable, for different machines, probably always "unknown"?